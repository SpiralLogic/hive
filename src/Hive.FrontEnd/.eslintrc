{
  "plugins": [
    "@typescript-eslint",
    "react-hooks",
    "testing-library",
    "jest-dom",
    "unicorn"
  ],
  "env": {
    "browser": true,
    "es6": true
  },
  "extends": [
    "kentcdodds",
    "plugin:jest/recommended",
    "plugin:jest-dom/recommended",
    "plugin:prettier/recommended",
    "plugin:testing-library/react",
    "plugin:testing-library/dom",
    "prettier",
    "plugin:unicorn/recommended",
    "preact"
  ],
  "rules": {
    "unicorn/no-null": "off",
    "no-unused-vars": "off",
    "@typescript-eslint/no-unused-vars": [
      "error"
    ]
  },
  "globals": {
    "Atomics": "readonly",
    "SharedArrayBuffer": "readonly",
    "h": true
  },
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "project": "./tsconfig.json",
    "jsx": true
  },
  "overrides": [
    {
      "files": [
        "**/*.jsx",
        "**/*.tsx"
      ],
      "rules": {
        "unicorn/filename-case": "off"
      }
    }
  ],
  "settings": {
    "react": {
      "pragma": "h"
    },
    "propWrapperFunctions": [
      "forbidExtraProps",
      {
        "property": "freeze",
        "object": "Object"
      }
    ],
    "linkComponents": [
      "Hyperlink",
      {
        "name": "Link",
        "linkAttribute": "to"
      }
    ]
  },
  "ignorePatterns": [
    "/node_modules/*/**",
    "/*.config.js",
    "/src/public/*/**",
    "**/__coverage__/**",
    "/src/index.tsx",
    "/src/index.html",
    "/build.js",
    "/inject.shim.js"
  ]
}
