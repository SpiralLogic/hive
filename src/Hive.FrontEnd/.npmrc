node-version=20.11.1