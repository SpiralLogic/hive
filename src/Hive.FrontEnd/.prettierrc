{
  "singleQuote": true,
  "printWidth": 110,
  "proseWrap": "always",
  "bracketSameLine": true,
  "endOfLine": "auto",
  "trailingComma": "es5",
  "tabWidth": 2
}
