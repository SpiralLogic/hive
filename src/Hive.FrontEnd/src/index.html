<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="description" content="Hive online board game">
  <title>Hive</title>
  [css]
  [scripts]    
</head>
</html>
