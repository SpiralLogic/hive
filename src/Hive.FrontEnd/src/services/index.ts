export * from './hextille-builder';
export * from './hive-dispatcher';
export * from './server-connection';
