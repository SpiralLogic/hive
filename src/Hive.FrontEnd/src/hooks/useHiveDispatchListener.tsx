import { HiveDispatcher, HiveEventListener, HiveIntent } from '../services';
import { useContext, useEffect } from 'preact/hooks';
import { createContext } from 'preact';

export const dispatcher = new HiveDispatcher();
export const Dispatcher = createContext(dispatcher);
export const useHiveDispatchListener = <T extends HiveIntent>(
  type: T['type'],
  listener: HiveEventListener<T>
): void => {
  const dispatcher = useContext(Dispatcher);
  useEffect(() => {
    return dispatcher.add<T>(type, listener);
  }, [type, listener]);
};
