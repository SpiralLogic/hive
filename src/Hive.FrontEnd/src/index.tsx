import { h, render } from 'preact';
import App from './components/App';
import { serverConnectionFactory } from './services';
import creatures from './svg/creatures.svg?raw';

import GameEngine from './services/game-engine';

document.body.insertAdjacentHTML('beforeend', `${creatures}`);
document.title = import.meta.env.HIVE_APP_TITLE;

const [, , gameId, currentPlayer] = window.location.pathname.split('/');
const engine = new GameEngine({ gameId, currentPlayer });
const properties = {
  engine,
  connectionFactory: serverConnectionFactory,
};

render(h(App, properties), document.body);

export { removeOtherPlayerMoves } from './utilities/hextille';
export { cellKey } from './utilities/hextille';
console.log(import.meta.env);
