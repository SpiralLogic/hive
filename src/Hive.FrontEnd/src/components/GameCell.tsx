import { FunctionComponent } from 'preact';
import { useEffect, useState } from 'preact/hooks';
import { HexCoordinates, Tile as TileType } from '../domain';
import { TileEvent } from '../services';
import { addHiveDispatchListener, dispatchHiveEvent } from '../utilities/dispatcher';
import { handleDragOver, handleKeyboardNav, isEnterOrSpace } from '../utilities/handlers';
import { useClassReducer } from '../utilities/class-reducer';
import Hexagon from './Hexagon';
const isValidMove = (validMoves: HexCoordinates[], coords: HexCoordinates) =>
  validMoves.some((dest) => dest.q == coords.q && dest.r == coords.r);

type Props = { coords: HexCoordinates; hidden?: boolean };
const GameCell: FunctionComponent<Props> = (props) => {
  const { coords, children, hidden } = props;
  const [classes, setClasses] = useClassReducer('hide');
  const [selectedTile, setSelectedTile] = useState<TileType | null>(null);

  useEffect(() => setClasses({ type: hidden ? 'add' : 'remove', classes: ['hide'] }), [hidden]);

  addHiveDispatchListener<TileEvent>('tileDeselected', () => {
    setSelectedTile(null);
    setClasses({ type: 'remove', classes: ['active', 'can-drop', 'no-drop'] });
  });

  addHiveDispatchListener<TileEvent>('tileSelected', (intent) => {
    if (!isValidMove(intent.tile.moves, coords)) {
      setClasses({ type: 'add', classes: ['no-drop'] });
    } else {
      setSelectedTile(intent.tile);
      setClasses({ type: 'add', classes: ['can-drop'] });
    }
  });

  addHiveDispatchListener<TileEvent>('tileDropped', () => {
    if (classes.includes('active') && selectedTile && isValidMove(selectedTile.moves, coords)) {
      dispatchHiveEvent({
        type: 'move',
        move: { coords, tileId: selectedTile.id },
      });
    }
  });

  const handleDragLeave = (event: DragEvent) => {
    event.stopPropagation();
    setClasses({ type: 'remove', classes: ['active'] });
  };

  const handleDragEnter = () => {
    setClasses({ type: 'add', classes: ['active'] });
  };

  const handleClick = (event: UIEvent) => {
    if (!(selectedTile && isValidMove(selectedTile.moves, coords))) return;
    event.stopPropagation();
    dispatchHiveEvent({ type: 'tileClear', tile: selectedTile });
    dispatchHiveEvent({ type: 'move', move: { coords, tileId: selectedTile.id } });
  };

  const handleKeydown = (e: KeyboardEvent) => {
    if (handleKeyboardNav(e) || !isEnterOrSpace(e)) return;
    handleClick(e);
  };

  const attributes = {
    class: classes || undefined,
    tabindex: selectedTile && isValidMove(selectedTile.moves, coords) ? 0 : undefined,
    role: hidden ? 'none' : 'cell',
  };

  const handlers = {
    ondragover: handleDragOver,
    ondragleave: handleDragLeave,
    ondragenter: handleDragEnter,
    onmouseenter: handleDragEnter,
    onmouseleave: handleDragLeave,
    onclick: handleClick,
    onkeydown: handleKeydown,
  };
  return (
    <Hexagon hidden={hidden} {...attributes} {...handlers}>
      {children}
    </Hexagon>
  );
};
GameCell.displayName = 'Cell';
export default GameCell;
