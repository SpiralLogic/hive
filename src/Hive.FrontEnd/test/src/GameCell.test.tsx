import { fireEvent, render, screen } from '@testing-library/preact';
import userEvent from '@testing-library/user-event';
import { waitFor } from '@testing-library/dom';
import { ComponentProps } from 'preact';
import GameCell from '../../src/components/GameCell';
import { simulateEvent } from '../helpers';
import {
  createCellCanDrop,
  createCellMovableTile,
  createCellNoDrop,
  createCellWithNoTile,
  createCellWithTile,
  createCellWithTileAndDrop,
  createCellWithTileAndHistoricalMove,
  createCellWithTileNoDrop,
  createMoveListener,
  movingTile,
} from '../fixtures/game-cell.fixtures';
import { HiveDispatcher } from '../../src/services';
import { Dispatcher } from '../../src/hooks/useHiveDispatchListener';

const setUp = (...tileCreationFns: Array<() => ComponentProps<typeof GameCell>>) => {
  const dispatcher = new HiveDispatcher();
  render(
    <Dispatcher.Provider value={dispatcher}>
      {tileCreationFns.map((factory) => (
        <GameCell key={factory} {...factory()} />
      ))}
    </Dispatcher.Provider>
  );
  return dispatcher;
};

describe('<GameCell>', () => {
  it(`top tile is rendered when tiles are all empty`, async () => {
    setUp(createCellWithTile);
    expect(screen.getByTitle(/fly/)).toBeInTheDocument();
  });

  it(`shows cells with historical moves`, async () => {
    setUp(createCellWithTileAndHistoricalMove);

    expect(screen.getByRole('cell')).toHaveClass('historical');
  });

  it('allows drop on dragover', async () => {
    setUp(createCellWithTile);
    const preventDefault = simulateEvent(screen.getByRole('cell'), 'dragover');
    expect(preventDefault).toHaveBeenCalledWith();
  });

  it('is available on drag start', async () => {
    setUp(createCellMovableTile, createCellWithTileAndDrop, createCellCanDrop);
    await userEvent.click(screen.getByTitle(/tilecanmove/));
    const cells = await screen.findAllByRole('cell');
    await Promise.all(cells.slice(1).map((c) => waitFor(() => expect(c).toHaveClass('can-drop'))));
  });

  it('is active on tile drag enter', async () => {
    setUp(createCellMovableTile, createCellWithTileAndDrop, createCellCanDrop);

    await userEvent.click(screen.getByTitle(/tilecanmove/));

    for (const c of screen.getAllByRole('cell')) fireEvent.dragEnter(c);

    await Promise.all(screen.getAllByRole('cell').map((c) => waitFor(() => expect(c).toHaveClass('active'))));
  });

  it('is no longer active on tile drag leave', async () => {
    setUp(createCellMovableTile, createCellWithTileAndDrop, createCellCanDrop);

    await userEvent.click(screen.getByTitle(/tilecanmove/));

    for (const c of screen.getAllByRole('cell')) fireEvent.dragEnter(c);
    for (const c of screen.getAllByRole('cell')) fireEvent.dragLeave(c);
    await Promise.all(
      screen.getAllByRole('cell').map((c) => waitFor(() => expect(c).not.toHaveClass('active')))
    );
  });

  it('calls move when cell is valid and active for tile being dragged', async () => {
    const dispatcher = setUp(createCellMovableTile, createCellWithTileAndDrop, createCellCanDrop);
    const moveEvents = createMoveListener(dispatcher);

    await userEvent.click(screen.getByTitle(/tilecanmove/));

    for (const c of screen.getAllByRole('cell')) fireEvent.dragEnter(c);
    dispatcher.dispatch({
      type: 'tileDropped',
      tile: movingTile,
    });

    expect(moveEvents).toStrictEqual(
      expect.arrayContaining([
        {
          type: 'move',
          move: { tileId: 2, coords: { q: 0, r: 0 } },
        },
        {
          type: 'move',
          move: { tileId: 2, coords: { q: 2, r: 2 } },
        },
      ])
    );
  });

  it(`doesn't call move tile on drop when cell doesn't allow drop`, async () => {
    const dispatcher = setUp(createCellMovableTile, createCellWithTile, createCellNoDrop);
    const moveEvents = createMoveListener(dispatcher);

    await userEvent.click(screen.getByTitle(/tilecanmove/));

    dispatcher.dispatch({
      type: 'tileDropped',
      tile: movingTile,
    });

    expect(moveEvents).toStrictEqual([]);
  });

  it(`don't call move tile on drop for invalid cells`, async () => {
    const dispatcher = setUp(createCellWithTileNoDrop, createCellNoDrop);
    const moveEvents = createMoveListener(dispatcher);

    for (const c of screen.getAllByRole('cell')) fireEvent.dragEnter(c);
    expect(moveEvents).toStrictEqual([]);
  });

  it('removes active classes on drag leave', async () => {
    setUp(createCellMovableTile, createCellWithTileNoDrop, createCellNoDrop);

    await userEvent.click(screen.getByTitle(/tilecanmove/));

    for (const c of screen.getAllByRole('cell')) fireEvent.dragEnter(c);
    for (const c of screen.getAllByRole('cell')) fireEvent.dragLeave(c);
    await Promise.all(
      screen.getAllByRole('cell').map((c) => waitFor(() => expect(c).not.toHaveClass('active')))
    );
  });

  it('removes active and can-drop classes on drop', async () => {
    const dispatcher = setUp(
      createCellMovableTile,
      createCellWithTile,
      createCellWithTileNoDrop,
      createCellNoDrop
    );

    await userEvent.click(screen.getByTitle(/tilecanmove/));

    for (const c of screen.getAllByRole('cell')) fireEvent.dragEnter(c);

    dispatcher.dispatch({
      type: 'tileDropped',
      tile: movingTile,
    });

    dispatcher.dispatch({
      type: 'tileDeselected',
      tile: movingTile,
    });

    dispatcher.dispatch({
      type: 'tileDeselected',
      tile: movingTile,
    });

    for (const c of screen.getAllByRole('cell')) expect(c).not.toHaveClass('active');
    for (const c of screen.getAllByRole('cell')) expect(c).not.toHaveClass('can-drop');
  });

  it(`doesn't stop event propagation when occupied cell has no active tile '`, async () => {
    const dispatcher = setUp(createCellWithTileNoDrop);
    const moveEvents = createMoveListener(dispatcher);

    await Promise.all(screen.getAllByRole('cell').map((c) => userEvent.click(c)));

    expect(moveEvents).toStrictEqual([]);
  });

  it(`makes a move when clicked with active tile`, async () => {
    const dispatcher = setUp(createCellMovableTile, createCellCanDrop);
    const moveEvents = createMoveListener(dispatcher);

    await userEvent.click(screen.getByTitle(/tilecanmove/));
    const cells = await screen.findAllByRole('cell');
    await userEvent.click(cells[1]);
    expect(moveEvents).toStrictEqual(
      expect.arrayContaining([
        {
          type: 'move',
          move: { tileId: 2, coords: { q: 0, r: 0 } },
        },
      ])
    );
  });

  it(`shouldn't move with no active tile on cell click`, async () => {
    const dispatcher = setUp(createCellNoDrop);
    const moveEvents = createMoveListener(dispatcher);

    await Promise.all(screen.getAllByRole('cell').map((c) => userEvent.click(c)));

    expect(moveEvents).toStrictEqual([]);
  });

  it(`shouldn't move with invalid tile on cell click`, async () => {
    const dispatcher = setUp(createCellMovableTile, createCellNoDrop);

    const moveEvents = createMoveListener(dispatcher);

    await userEvent.click(screen.getByTitle(/tilecanmove/));
    const cells = await screen.findAllByRole('cell');
    await userEvent.click(cells[1]);
    expect(moveEvents).toStrictEqual([]);
  });

  it(`fires emit event when pressing enter`, async () => {
    const dispatcher = setUp(createCellMovableTile, createCellCanDrop);

    const moveEvents = createMoveListener(dispatcher);
    const tileCanMove = screen.getByTitle(/tilecanmove/);

    await userEvent.click(tileCanMove);
    await userEvent.tab();
    await userEvent.keyboard('{Enter}');

    expect(moveEvents).toStrictEqual(expect.arrayContaining([expect.objectContaining({ type: 'move' })]));
  });

  it(`dispatches move event when pressing space`, async () => {
    const dispatcher = setUp(createCellMovableTile, createCellCanDrop);

    const moveEvents = createMoveListener(dispatcher);

    await userEvent.click(screen.getByTitle(/tilecanmove/));
    await userEvent.tab();
    await userEvent.keyboard(' ');

    expect(moveEvents).toStrictEqual(expect.arrayContaining([expect.objectContaining({ type: 'move' })]));
  });

  it.each([
    '{backspace}',
    '{esc}',
    '{del}',
    '{selectall}',
    '{arrowleft}',
    '{arrowright}',
    '{arrowup}',
    '{arrowdown}',
    '{home}',
    '{end}',
    '{shift}',
    '{ctrl}',
    '{alt}',
    '{meta}',
    '{capslock}',
    ...Array.from({ length: 79 })
      .map((c, index) => String.fromCodePoint(index + 48))
      .map((key) => key.replaceAll('{', '{{').replaceAll('[', '[[')),
  ])(`doesn't start drag event for key %s`, async (key) => {
    const dispatcher = setUp(createCellMovableTile, createCellCanDrop);

    const moveEvents = createMoveListener(dispatcher);

    await userEvent.click(screen.getByTitle(/tilecanmove/));
    const cells = await screen.findAllByRole('cell');
    cells[1].focus();
    await userEvent.keyboard(key);

    expect(moveEvents).toStrictEqual([]);
  });

  it('renders cell with tile', () => {
    setUp(createCellWithTile);
    expect(screen.getByRole('cell')).toMatchSnapshot();
  });

  it('renders cell with no tile', () => {
    setUp(createCellWithNoTile);
    expect(screen.getByRole('cell')).toMatchSnapshot();
  });
});
