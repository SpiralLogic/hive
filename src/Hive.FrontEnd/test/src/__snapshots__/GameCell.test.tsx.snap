// Vitest Snapshot v1

exports[`<GameCell> snapshots > matches cell with no tile 1`] = `
<div
  role="cell"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
  </svg>
</div>
`;

exports[`<GameCell> snapshots > matches cell with tile 1`] = `
<div
  role="cell"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <use
      href="#hex"
    />
  </svg>
  <div
    class="player1 hex tile fly"
    draggable="false"
    title="Player-1 fly"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        href="#hex"
      />
      <use
        href="#fly"
      />
    </svg>
  </div>
</div>
`;
